pricers.py

```diff
@@ -0,0 +1,321 @@
+            x: state tensor with shape [B, M, N, d], [B, M, N, 2d] under SV model, [B, M, N, 3d] under SV model for path dependent product
+            u_c: function embedding tensor [B, M, N, k_1]
+            u_p: other parameters tensor [B, M, N, k_2]
+            return: the out put function value with shape [B, M, N, 1]
+        """
+        t, x, u_hat = inputs
+        if type(self.no_net) == DeepONet:
+            y = self.no_net((t, x, u_hat))
+        # print(t.shape, x.shape, u.shape)
+        else:
+            u_c, u_p = self.sde.split_uhat(u_hat)
+            # print(u_c.shape, u_p.shape)
+            y = self.no_net((t, x, u_c, u_p))
+        return y
+        
+    def loss_fn(self, data: Tuple[tf.Tensor], training=None):
+        raise NotImplementedError
+
+    @tf.function
+    def train_step(self, inputs: Tuple[tf.Tensor]) -> dict:
+        raise NotImplementedError
+
+    def drift_bsde(self, t: tf.Tensor, x: tf.Tensor, y: tf.Tensor, u_hat: tf.Tensor) -> tf.Tensor: 
+        """
+        the driver term of the BSDE, which is determined by the SDE
+        """
+        raise NotImplementedError
+
+    def diffusion_bsde(self, t: tf.Tensor, x: tf.Tensor, grad: tf.Tensor, dw: tf.Tensor, u_hat: tf.Tensor) -> tf.Tensor:
+        """
+        the diffusion erm of the BSDE, which is determined by the SDE
+        """
+        raise NotImplementedError
+
+    def payoff_func(self, t: tf.Tensor, x: tf.Tensor, u_hat: tf.Tensor) -> tf.Tensor:
+        """
+        The payoff function of the time horizon and paths based on the given time tensor and state tensor
+        """
+        raise NotImplementedError
+
+
+class MarkovianPricer(BaseBSDEPricer):
+    """
+    This is the class to construct the tain step of the BSDE loss function, the data is generated from three
+    external classes: the sde class which yield the data of input parameters. the option class which yields the
+    input function data and give the payoff function and exact option price.
+    """
+
+    def __init__(self, sde, option, config):
+        super(MarkovianPricer, self).__init__(sde, option, config)
+
+    def loss_fn(self, data: Tuple[tf.Tensor], training=None) -> Tuple[tf.Tensor]:
+        """
+        :param t: time B x M x (T-1) x 1
+        :param x: state B x M x (T-1) x D
+        :param  u: B x K ->(repeat) B x M x (T-1) x K
+        :return: value (B, M, (T-1), 1), gradient_x (B, M, (T-1), D)
+        """
+        t, x, dw, u_hat = data
+        steps = self.eqn_config.time_steps
+        loss_interior = 0.0
+        u_now = u_hat[:, :, :-1, :]
+        x_now = x[:, :, :-1, :]
+        t_now = t[:, :, :-1, :]
+        with tf.GradientTape(watch_accessed_variables=False) as tape:
+            tape.watch(x)
+            f = self.call((t, x, u_hat))
+            grad = tape.gradient(f, x) 
+        f_now = f[:, :, :-1, :]
+        f_pls = f[:, :, 1:, :]
+        grad = grad[:,:,:-1,:]
+        for n in range(steps-1):
```

Comment:
@gengem777 This is an important change. Let's make it a separate function, and also rewrite the loop here -- we could use a tensor-wise operation to replace it (if memory not exploding).

Reply:
Also, here operations are intense with high-dim tensors, so lets add shape info for each single line in this function -- later when we debug this would be extremely useful.